src/sbvr-api/translations.ts

```diff
@@ -0,0 +1,215 @@
+	});
+};
+
+export const translateAbstractSqlModel = (
+	fromAbstractSqlModel: AbstractSqlModel,
+	toAbstractSqlModel: AbstractSqlModel,
+	fromVersion: string,
+	toVersion: string,
+	definitions: _.Dictionary<
+		| (Definition & { $toResource?: string })
+		| _.Dictionary<string | ReferencedFieldNode>
+	> = {},
+): _.Dictionary<string> => {
+	const resourceRenames: _.Dictionary<string> = {};
+
+	// TODO: why?
+	fromAbstractSqlModel.rules = toAbstractSqlModel.rules;
+
+	const fromKeys = Object.keys(fromAbstractSqlModel.tables);
+	const nonexistentTables = _.difference(Object.keys(definitions), fromKeys);
+	if (nonexistentTables.length > 0) {
+		throw new Error(
+			`Tried to define non-existent resources: '${nonexistentTables.join(
+				', ',
+			)}'`,
+		);
+	}
+	_.forEach(toAbstractSqlModel.synonyms, (canonicalForm, synonym) => {
+		// Don't double alias
+		if (synonym.includes('$')) {
+			fromAbstractSqlModel.synonyms[synonym] = canonicalForm;
+		} else {
+			fromAbstractSqlModel.synonyms[
+				`${synonym}$${toVersion}`
+			] = `${canonicalForm}$${toVersion}`;
+		}
+	});
+	const relationships = _.cloneDeep(toAbstractSqlModel.relationships);
+	namespaceRelationships(relationships, toVersion);
+	_.forEach(relationships, (relationship, key) => {
+		// Don't double alias
+		if (!key.includes('$')) {
+			key = `${key}$${toVersion}`;
+		}
+		fromAbstractSqlModel.relationships[key] = relationship;
+	});
+
+	// TODO: We also need to keep the original relationship refs to non $version resources
+	// Also alias for ourselves to allow explicit referencing
+	const aliasedFromRelationships = _.cloneDeep(
+		fromAbstractSqlModel.relationships,
+	);
+	namespaceRelationships(aliasedFromRelationships as Relationship, fromVersion);
+	_.forEach(aliasedFromRelationships, (relationship, key) => {
+		// Don't double alias
+		if (!key.includes('$')) {
+			key = `${key}$${fromVersion}`;
+			fromAbstractSqlModel.relationships[key] = relationship;
+		}
+	});
+
+	_.forEach(toAbstractSqlModel.tables, (table, key) => {
+		// Don't double alias
+		if (!key.includes('$')) {
+			key = `${key}$${toVersion}`;
+		}
+		fromAbstractSqlModel.tables[key] = _.cloneDeep(table);
+	});
+
+	fromKeys.forEach((key) => {
+		const definition = definitions[key];
+		const table = fromAbstractSqlModel.tables[key];
+		if (definition) {
+			const hasToResource = typeof definition.$toResource === 'string';
+			if (hasToResource) {
+				resourceRenames[key] = `${definition.$toResource}`;
+			}
+			const toResource = hasToResource
+				? (definition.$toResource as string)
+				: `${key}$${toVersion}`;
+			// TODO: Should this use the toAbstractSqlModel?
+			const toTable = fromAbstractSqlModel.tables[toResource];
+			if (!toTable) {
+				if (hasToResource) {
+					throw new Error(`Unknown $toResource: '${toResource}'`);
+				} else {
+					throw new Error(`Missing $toResource: '${toResource}'`);
+				}
+			}
+			table.modifyFields = _.cloneDeep(toTable.modifyFields ?? toTable.fields);
```

Comment:
I think that I couldn't spot a place that we might me modifying `modifyFields` or `fields` and also tried to change their typings to ReadonlyArray, but TS didn't complain :thinking: 
No big deal though, since it sounds safer to completely isolate the models.

Reply:
Yeah I don't think the `cloneDeep` is strictly necessary based upon pinejs code as it stands today but I think it's worth keeping the different models isolated or there could be very weird and hard to debug bugs that occur due to editing eg `v5.modifyFields` and seeing `v6.fields` change, or vice-versa, or some other combo, which way it's not something I want to debug :P